src/coreclr/vm/methodtable.h

```diff
@@ -296,31 +300,39 @@ struct MethodTableWriteableData
         // TO BE UPDATED IN ORDER TO ENSURE THAT METHODTABLES DUPLICATED FOR GENERIC INSTANTIATIONS
         // CARRY THE CORRECT INITIAL FLAGS.
 
-        enum_flag_Unrestored                = 0x00000004,
-        enum_flag_HasApproxParent           = 0x00000010,
-        enum_flag_UnrestoredTypeKey         = 0x00000020,
-        enum_flag_IsNotFullyLoaded          = 0x00000040,
-        enum_flag_DependenciesLoaded        = 0x00000080,     // class and all dependencies loaded up to CLASS_LOADED_BUT_NOT_VERIFIED
+        enum_flag_CanCompareBitsOrUseFastGetHashCode       = 0x0001,     // Is any field type or sub field type overrode Equals or GetHashCode
+        enum_flag_HasCheckedCanCompareBitsOrUseFastGetHashCode   = 0x0002,  // Whether we have checked the overridden Equals or GetHashCode
 
-        // enum_unused                      = 0x00000100,
+        enum_flag_Unrestored                = 0x0004,
+        enum_flag_HasApproxParent           = 0x0010,
+        enum_flag_UnrestoredTypeKey         = 0x0020,
+        enum_flag_IsNotFullyLoaded          = 0x0040,
+        enum_flag_DependenciesLoaded        = 0x0080,     // class and all dependencies loaded up to CLASS_LOADED_BUT_NOT_VERIFIED
 
-        enum_flag_CanCompareBitsOrUseFastGetHashCode       = 0x00000200,     // Is any field type or sub field type overrode Equals or GetHashCode
-        enum_flag_HasCheckedCanCompareBitsOrUseFastGetHashCode   = 0x00000400,  // Whether we have checked the overridden Equals or GetHashCode
-
-        // enum_unused                      = 0x00010000,
-        // enum_unused                      = 0x00020000,
-        // enum_unused                      = 0x00040000,
-        // enum_unused                      = 0x00080000,        // enum_unused                      = 0x0010000,
-        // enum_unused                      = 0x0020000,
-        // enum_unused                      = 0x0040000,
-        // enum_unused                      = 0x0080000,
+        // enum_unused                      = 0x0100,
+        // enum_unused                      = 0x0200,
+        // enum_unused                      = 0x0400,
+        // enum_unused                      = 0x0800,
+        // enum_unused                      = 0x1000,
+        // enum_unused                      = 0x2000,
 
 #ifdef _DEBUG
-        enum_flag_ParentMethodTablePointerValid =  0x40000000,
-        enum_flag_HasInjectedInterfaceDuplicates = 0x80000000,
+        enum_flag_ParentMethodTablePointerValid =  0x4000,
+        enum_flag_HasInjectedInterfaceDuplicates = 0x8000,
 #endif
     };
-    DWORD      m_dwFlags;                  // Lot of empty bits here.
+    union
+    {
+        DWORD      m_dwFlags;                  // Lot of empty bits here.
+        struct
+        {
+            uint16_t m_loFlags;
+            int16_t m_offsetToNonVirtualSlots;
+        };
+    };
+    
+
+    PTR_Module m_pLoaderModule;
```

Comment:
Is there a reason for `m_pLoaderModule` to be here and not in the MethodTable?

Reply:
No strong reason, I was trying to avoid growing the `MethodTable` structure, as growth has a (very) small impact on v-table slots due to the limited range of a mov instruction on x64, but its not a critical issue.

Effectively, there is a small size optimization to be had in generated code for virtual function access where if the offset of the vtable chunk index is less than 256 bytes from the start of the `MethodTable`, we can use a 3 byte instruction instead of a 6 byte instruction. It's not significant for throughput perf, but historically I remember seeing some not completely negligible impact on code size for ngen images when adjusting the size of `MethodTable`. However, of course, ngen is no longer a thing, so perhaps I shouldn't worry about that, and looking at the data, I see that it would only matter for virtual table slots 25-32 if we moved two pointer sized values down from `MethodTableWriteableData` to `MethodTable`, which probably isn't significant at all. (Possibly my memories are from the *really* before times, when we didn't have vtable chunks, and this change would have effected vtable slots 7 and 8, which are fairly common in code bases)

From what I can see the `m_pLoaderModule`'s only real perf sensitive use is as part of statics lookup, and I'm planning to make that simpler in a followon checkin, which is why I didn't keep it at the `MethodTable` level, but you are correct, it probably would be fine as a field of MethodTable (along with the `m_hExposedClassObject` field which also has most of the same properties) I'd love to hear if anyone else has an opinion here.

Reply:
I think that the state in this PR is fine. We can fine-tune it in follow up changes.